### PR TITLE
v2

### PR DESCRIPTION
**Pull request by yonishelach**

### CODE REVIEW

File: src/hub_functions/structured_data_generator.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    if key in os.environ and base in os.environ:
+        return True
+    # Check if mlrun is installed:
+    try:
+        import mlrun
+    except ModuleNotFoundError:
+        raise EnvironmentError(
+            f"One or more of the OpenAI required environment variables ('{key}', '{base}') are missing."
+            f"Please set them as environment variables or install mlrun (`pip install mlrun`)"
+            f"and set them as project secrets using `projecy.set_secrets`."
```
Comment (guy1992l) [inline src/hub_functions/structured_data_generator.py:34]:
```suggestion
            f"and set them as project secrets using `project.set_secrets`."
```

File: src/hub_functions/structured_data_generator.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import os
+
+import tqdm
+from langchain.chat_models import ChatOpenAI
+
+
+def _set_openai_secrets() -> bool:
+    key = "OPENAI_API_KEY"
+    base = "OPENAI_API_BASE"
+    # Check if the key is already in the environment variables:
```
Comment (guy1992l) [inline src/hub_functions/structured_data_generator.py:24]:
```suggestion
    
    # Check if the key is already in the environment variables:
```

File: src/hub_functions/structured_data_generator.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    if key in os.environ and base in os.environ:
+        return True
+    # Check if mlrun is installed:
```
Comment (guy1992l) [inline src/hub_functions/structured_data_generator.py:27]:
```suggestion
    
    # Check if mlrun is installed:
```

File: src/hub_functions/text_to_audio_generator.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+            audio = np.vstack(
+                [np.concatenate(audio_pieces[speaker]) for speaker in speakers]
+            ).astype(dtype=np.float32)
+            # Resample:
+            audio = torch.from_numpy(audio)
+            audio = resampler(audio)
+            # Save to audio file:
+            audio_file = output_directory / f"{text_file.stem}.{file_format}"
+
```
Comment (guy1992l) [inline src/hub_functions/text_to_audio_generator.py:183]:
```suggestion
```

File: src/hub_functions/text_to_audio_generator.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+                        silence = np.zeros_like(audio)
+                        for speaker in audio_pieces.keys():
+                            if speaker == current_speaker:
+                                audio_pieces[speaker] += [audio, gap_between_speakers]
+                            else:
+                                audio_pieces[speaker] += [silence, gap_between_speakers]
+                    else:
+                        audio_pieces["all"] += [audio, gap_between_speakers]
+            # Construct a single audio array from all the pieces and channels:
+
```
Comment (guy1992l) [inline src/hub_functions/text_to_audio_generator.py:174]:
```suggestion
```

File: src/hub_functions/text_to_audio_generator.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+            audio_pieces = (
+                {speaker: [] for speaker in speakers}
+                if speaker_per_channel
+                else {"all": []}
+            )
+
+            # Generate audio per line:
+            for line in text.splitlines():
+                # Validate line is in correct speaker format:
+
```
Comment (guy1992l) [inline src/hub_functions/text_to_audio_generator.py:145]:
```suggestion
```

File: src/hub_functions/text_to_audio_generator.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+            # Read text:
+            with open(text_file, "r") as fp:
+                text = fp.read()
+            # Prepare a holder for all the generated pieces (if per channel each speaker will have its own):
+            audio_pieces = (
+                {speaker: [] for speaker in speakers}
+                if speaker_per_channel
+                else {"all": []}
+            )
+
```
Comment (guy1992l) [inline src/hub_functions/text_to_audio_generator.py:141]:
```suggestion
```

File: src/hub_functions/text_to_audio_generator.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    output_directory = pathlib.Path(output_directory)
+    if not output_directory.exists():
+        output_directory.mkdir(exist_ok=True, parents=True)
+
+    # Start generating audio:
+    # Go over the audio files and transcribe:
+    for text_file in tqdm.tqdm(
+        text_files, desc="Generating", unit="file", disable=not verbose
+    ):
+
```
Comment (guy1992l) [inline src/hub_functions/text_to_audio_generator.py:123]:
```suggestion
```

File: src/hub_functions/text_to_audio_generator.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+    # Create the output directory:
+    if output_directory is None:
+        output_directory = tempfile.mkdtemp()
+    output_directory = pathlib.Path(output_directory)
+    if not output_directory.exists():
+        output_directory.mkdir(exist_ok=True, parents=True)
+
+    # Start generating audio:
+    # Go over the audio files and transcribe:
```
Comment (guy1992l) [inline src/hub_functions/text_to_audio_generator.py:119]:
```suggestion
```

File: src/hub_functions/text_to_audio_generator.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    :param verbose:             Whether to print the progress of the generation.
+    :param bits_per_sample:     Changes the bit depth for the supported formats.
+                                Supported only in "wav" or "flac" formats.
+
+    :returns:                   A tuple of:
+                                - The output directory path.
+                                - The generated audio files dataframe.
+                                - The errors dictionary.
+    """
+
```
Comment (guy1992l) [inline src/hub_functions/text_to_audio_generator.py:69]:
```suggestion
```

File: src/hub_functions/text_to_audio_generator.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+    global _LOGGER
+    _LOGGER = _get_logger()
+    # Get the input text files to turn to audio:
```
Comment (guy1992l) [inline src/hub_functions/text_to_audio_generator.py:72]:
```suggestion
    
    # Get the input text files to turn to audio:
```

File: src/hub_functions/text_to_audio_generator.py [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+            audio_file = output_directory / f"{text_file.stem}.{file_format}"
+
+            torchaudio.save(
+                uri=str(audio_file),
+                src=audio,
+                sample_rate=sample_rate,
+                format=file_format,
+                bits_per_sample=bits_per_sample,
+            )
+
```
Comment (guy1992l) [inline src/hub_functions/text_to_audio_generator.py:191]:
```suggestion
```

File: src/workflows/calls_generation.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    min_time: int,
+    max_time: int,
+    from_date: str,
+    to_date: str,
+    from_time: str,
+    to_time: str,
+    num_clients: int,
+    num_agents: int,
+    generate_clients_and_agents: bool = True,
+):
```
Comment (guy1992l) [inline src/workflows/calls_generation.py:38]: where all the doc strings went? I'm sure I wrote them, weird 😕

File: src/workflows/calls_analysis.py [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+    batch: str,
+    calls_audio_files: str,
+    transcribe_model: str,
+    translate_to_english: bool,
+    pii_recognition_model: str,
+    pii_recognition_entities: List[str],
+    pii_recognition_entity_operator_map: List[str],
+    question_answering_model: str,
+    batch_size: int = 2,
+):
```
Comment (guy1992l) [inline src/workflows/calls_analysis.py:109]: doc strings?

**Review by guy1992l — CHANGES_REQUESTED** [review]
no need for `translate.py`. Regarding the doc strings, let's leave it for now, once you mvoe to work with the hub, I'll sort it out

File: src/workflows/calls_generation.py (hunk 13 above)
Comment (yonishelach) [reply src/workflows/calls_generation.py:38]: I did not see them anywhere... maybe you wrote them locally?

**Review by guy1992l — APPROVED** [review]

**Pull request merged**